library/Translate/src/Juvix/FrontendContextualise.hs

```diff
@@ -31,42 +31,47 @@ data Error
 type Final f = Target.New f
 
 op ::
-  NonEmpty (NameSymbol.T, [Initial.TopLevel]) -> Either Error Target.FinalContext
+  NonEmpty (NameSymbol.T, [Initial.TopLevel]) -> IO (Either Error Target.FinalContext)
 op = contextualize
 
 contextualize ::
-  NonEmpty (NameSymbol.T, [Initial.TopLevel]) -> Either Error Target.FinalContext
-contextualize init =
-  case contextify init of
-    Left err -> Left (PathErr err)
-    Right (context, openList) ->
-      case Module.transformContext context openList of
-        Left err -> Left (ModuleErr err)
-        Right xs ->
-          case Infix.transformContext xs of
-            Left err -> Left (InfixErr err)
-            Right xs -> Right xs
+  NonEmpty (NameSymbol.T, [Initial.TopLevel]) -> IO (Either Error Target.FinalContext)
+contextualize init = do
+  ctx <- contextify init
+  case ctx of
+    Left err -> pure $ Left (PathErr err)
+    Right (context, openList) -> do
+      trans1 <- Module.transformContext context openList
+      case trans1 of
+        Left err -> pure $ Left (ModuleErr err)
+        Right xs -> do
+          infixed <- Infix.transformContext xs
+          case infixed of
+            Left err -> pure $ Left (InfixErr err)
+            Right xs -> pure $ Right xs
```

Comment:
is it possible to use `ExceptT` here too?

Reply:
It would be nice, I think I've opted out of that due to the error being differnet types all the way down